src/main/java/network/brightspots/rcv/CommonDataFormatReader.java

```diff
@@ -336,51 +261,183 @@ void parseCvrFile(List<CastVoteRecord> castVoteRecords)
-        ArrayList cvrSnapshots = (ArrayList) cvrObject.get("CVRSnapshot");
-        for (Object snapshotObject : cvrSnapshots) {
-          HashMap snapshot = (HashMap) snapshotObject;
-          if (!snapshot.get("@id").equals(currentSnapshotId)) {
-            continue;
-          }
 
-          // we found the current CVR snapshot so get rankings and create a new cvr
-          List<Pair<Integer, String>> rankings = parseRankingsFromSnapshot(snapshot, candidates);
-          CastVoteRecord newRecord =
-              new CastVoteRecord(computedCastVoteRecordId, ballotId, null, null, rankings);
-          castVoteRecords.add(newRecord);
+    ArrayList cvrs = (ArrayList) json.get("CVR");
+    for (Object cvrObject : cvrs) {
+      HashMap cvr = (HashMap) cvrObject;
+      HashMap cvrContest = getCvrContestJson(cvr, contestToTabulateId);
 
-          // provide some user feedback on the CVR count
-          if (castVoteRecords.size() % 50000 == 0) {
-            Logger.log(Level.INFO, "Parsed %d cast vote records.", castVoteRecords.size());
+      // parse selections
+      List<Pair<Integer, String>> rankings = new ArrayList<>();
+      ArrayList cvrContestSelections = (ArrayList) cvrContest.get("CVRContestSelection");
+      for (Object cvrContestSelectionObject : cvrContestSelections) {
+        HashMap cvrContestSelection = (HashMap) cvrContestSelectionObject;
+        String contestSelectionId = (String) cvrContestSelection.get("ContestSelectionId");
+        HashMap contestSelection = (HashMap) contestSelections.get(contestSelectionId);
+        String candidateId;
+        if (contestSelection.containsKey("IsWriteIn") && contestSelection.get("IsWriteIn")
+            .equals(BOOLEAN_TRUE)) {
+          // this is a write-in
+          candidateId = this.config.getUndeclaredWriteInLabel();
+        } else {
+          // lookup Candidate Name
+          ArrayList candidateIds = (ArrayList) contestSelection.get("CandidateIds");
+          if (candidateIds == null || candidateIds.size() == 0) {
+            Logger.severe("CandidateSelection \"%s\" has no CandidateIds!", contestSelectionId);
+            throw new CvrParseException();
+          }
+          if (candidateIds.size() > 1) {
+            Logger.warning("CandidateSelection \"%s\" has multiple CandidateIds.  "
+                + "Only the first one will be processed.", contestSelectionId);
+          }
+          String candidateObjectId = (String) candidateIds.get(0);
+          HashMap Candidate = (HashMap) candidates.get(candidateObjectId);
+          candidateId = (String) Candidate.get("Name");
+          if (candidateId.equals(config.getOvervoteLabel())) {
+            candidateId = Tabulator.EXPLICIT_OVERVOTE_LABEL;
+          }
+          if (!this.config.getCandidateCodeList().contains(candidateId)) {
+            Logger.severe("Unrecognized candidate found in CVR: %s", candidateId);
+            unrecognizedCandidateCounts.merge(candidateId, 1, Integer::sum);
+          }
```

Comment:
If you set candidateId to EXPLICIT_OVERVOTE_LABEL, isn't the check in line 385 guaranteed to fail? I think maybe 385 should be `else if` instead of `if`.

Reply:
Good catch!